src/libsystemd/sd-device/device-enumerator.c

```diff
@@ -278,16 +299,99 @@ static int device_compare(sd_device * const *_a, sd_device * const *_b) {
         return strcmp(devpath_a, devpath_b);
 }
 
+static int enumerator_sort_devices(sd_device_enumerator *enumerator) {
+        sd_device **devices;
+        sd_device *device;
+        size_t n = 0, m;
+        int r;
+
+        assert(enumerator);
+
+        devices = new(sd_device*, hashmap_size(enumerator->devices_by_syspath));
+        if (!devices)
+                return -ENOMEM;
+
+        if (!set_isempty(enumerator->prioritized_subsystem)) {
+                const char *syspath;
+
+                HASHMAP_FOREACH_KEY(device, syspath, enumerator->devices_by_syspath) {
+                        _cleanup_free_ char *p = NULL;
+                        const char *subsys;
+
+                        if (sd_device_get_subsystem(device, &subsys) < 0)
+                                continue;
+
+                        if (!set_contains(enumerator->prioritized_subsystem, subsys))
+                                continue;
+
+                        devices[n++] = sd_device_ref(device);
+
+                        for (;;) {
+                                _cleanup_free_ char *q = NULL;
+
+                                r = path_extract_directory(p ?: syspath, &q);
+                                if (r == -EADDRNOTAVAIL)
+                                        break;
+                                if (r < 0)
+                                        goto failed;
+
+                                device = hashmap_get(enumerator->devices_by_syspath, q);
+                                if (device)
+                                        devices[n++] = sd_device_ref(device);
+
+                                free_and_replace(p, q);
+                        }
+                }
+
+                for (size_t i = 0; i < n; i++) {
+                        r = sd_device_get_syspath(devices[i], &syspath);
+                        if (r < 0)
+                                goto failed;
+
+                        sd_device_unref(hashmap_remove(enumerator->devices_by_syspath, syspath));
+                }
+
+                typesafe_qsort(devices, n, device_compare);
+        }
+
+        m = n;
+        while ((device = hashmap_steal_first(enumerator->devices_by_syspath)))
+                devices[n++] = device;
+
+        typesafe_qsort(devices + m, n - m, device_compare);
+
+        enumerator->n_devices = n;
+        free_and_replace(enumerator->devices, devices);
+        return 0;
+
+failed:
+        for (size_t i = 0; i < n; i++)
+                sd_device_unref(devices[i]);
```

Comment:
i wonder if we should add a helper `device_unref_many(sd_device **array, size_t n);` that implements this loop. It appears to me we have that a couple of times.